### PR TITLE
Use generics in scheduling queue's heap

### PR DESCRIPTION
**Pull request by macsko**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

By introducing generics to Heap, scheduling queue can be simplified by removing type assertions. `podInfoKeyFunc` is also improved, by creating the `ObjectName` directly.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/internal/heap/heap.go [hunk 1]
```diff
@@ -115,7 +114,7 @@ func (h *data) Pop() interface{} {
 }
 
 // Peek is supposed to be called by heap.Peek only.
-func (h *data) Peek() interface{} {
+func (h *data[T]) Peek() interface{} {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:117]: why not return `T` instead of `interface{}`?

Comment (macsko) [reply pkg/scheduler/internal/heap/heap.go]: Good catch. `Peek()` can return `T` because it is not backed by [`container/heap`](https://pkg.go.dev/container/heap#Interface). Also, changed the comment not to be misleading.

File: pkg/scheduler/internal/heap/heap.go [hunk 2]
```diff
@@ -175,64 +169,63 @@ func (h *Heap) Delete(obj interface{}) error {
...
 	obj := heap.Pop(h.data)
 	if obj != nil {
 		if h.metricRecorder != nil {
 			h.metricRecorder.Dec()
 		}
-		return obj, nil
+		return obj.(T), nil
 	}
-	return nil, fmt.Errorf("object was removed from heap data")
+	var zero T
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:185]: ah, I thought golang already got https://github.com/golang/go/issues/61372..

File: pkg/scheduler/internal/heap/heap.go [hunk 3]
```diff
@@ -114,37 +113,35 @@ func (h *data) Pop() interface{} {
...
 }
 
 // Add inserts an item, and puts it in the queue. The item is updated if it
 // already exists.
-func (h *Heap) Add(obj interface{}) error {
-	key, err := h.data.keyFunc(obj)
-	if err != nil {
-		return cache.KeyError{Obj: obj, Err: err}
-	}
+func (h *Heap[T]) Add(obj T) error {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:138]: It never returns an error.

File: pkg/scheduler/internal/heap/heap.go [hunk 4]
```diff
@@ -81,28 +80,28 @@ func (h *data) Less(i, j int) bool {
...
 	item.index = i
 	item = h.items[h.queue[j]]
 	item.index = j
 }
 
-// Push is supposed to be called by heap.Push only.
-func (h *data) Push(kv interface{}) {
-	keyValue := kv.(*itemKeyValue)
+// Push is supposed to be called by container/heap.Push only.
+func (h *data[T]) Push(kv interface{}) {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:96]: `Push(kv *itemKeyValue[T])` would be better? (type check would work at compiling)

File: pkg/scheduler/internal/heap/heap.go [hunk 5]
```diff
@@ -81,28 +80,28 @@ func (h *data) Less(i, j int) bool {
...
 	n := len(h.queue)
-	h.items[keyValue.key] = &heapItem{keyValue.obj, n}
+	h.items[keyValue.key] = &heapItem[T]{keyValue.obj, n}
 	h.queue = append(h.queue, keyValue.key)
 }
 
-// Pop is supposed to be called by heap.Pop only.
-func (h *data) Pop() interface{} {
+// Pop is supposed to be called by container/heap.Pop only.
+func (h *data[T]) Pop() interface{} {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:104]: `Pop() T`?

File: pkg/scheduler/internal/heap/heap.go (hunk 4 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/heap/heap.go:96]: > supposed to be called by container/heap.Push only.

Okay, I missed this. 
So, Push and Pop use `interface()` to satisfy the interface of container/heap?

Comment (sanposhiho) [reply pkg/scheduler/internal/heap/heap.go:96]: Looks like Yes, and Golang has no generics support around container/heap yet. Okay, nvm.

File: pkg/scheduler/internal/heap/heap.go [hunk 6]
```diff
@@ -114,56 +113,50 @@ func (h *data) Pop() interface{} {
...
+		heap.Push(h.data, &itemKeyValue[T]{key, obj})
 		if h.metricRecorder != nil {
 			h.metricRecorder.Inc()
 		}
 	}
-	return nil
 }
 
 // Update is the same as Add in this implementation. When the item does not
 // exist, it is added.
```
Comment (sanposhiho) [inline pkg/scheduler/internal/heap/heap.go:152]: By the way, what's the reason behind we have both Add and Update while they're the same? Can we just have one of them?

File: pkg/scheduler/internal/heap/heap.go (hunk 4 above)
Comment (macsko) [reply pkg/scheduler/internal/heap/heap.go:96]: Exactly, `Push` and `Pop` have to use `interface{}`

**Review by sanposhiho — COMMENTED** [review]
It's great to see functions eliminate error from their return values.
One nit, others LGTM.

/lgtm

**Comment by sanposhiho** [comment]
/assign @kerthcet 

For approval.

File: pkg/scheduler/internal/heap/heap.go (hunk 6 above)
Comment (macsko) [reply pkg/scheduler/internal/heap/heap.go:152]: I would keep it for verbosity, to prevent making some kind of `Delete`-`Add` combo accidentally in the future.

Comment (sanposhiho) [reply pkg/scheduler/internal/heap/heap.go:152]: If your concern is that the naming would be confusing if we only have `Add` (or `Update`), can we name the function `Set`, or `AddOrUpdate` if we go more straightforward?

Comment (macsko) [reply pkg/scheduler/internal/heap/heap.go]: `AddOrUpdate` makes sense, renamed

**Review by sanposhiho — COMMENTED** [review]
Thanks for cleaning up Add / Update stuff too.
The linter CI is angry, but others LGTM.

**Comment by macsko** [comment]
Linter is passing now

**Review by sanposhiho — COMMENTED** [review]
/lgtm

File: pkg/scheduler/internal/queue/scheduling_queue_test.go [hunk 7]
```diff
@@ -2841,15 +2792,6 @@ func TestPodTimestamp(t *testing.T) {
...
 		},
-		{
-			name: "update two pod to activeQ and sort them by the timestamp",
-			operations: []operation{
-				updatePodActiveQ,
-				updatePodActiveQ,
-			},
-			operands: []*framework.QueuedPodInfo{pInfo2, pInfo1},
-			expected: []*framework.QueuedPodInfo{pInfo1, pInfo2},
-		},
```
Comment (gabesaba) [inline pkg/scheduler/internal/queue/scheduling_queue_test.go:2852]: why was this deleted?

**Review by gabesaba — COMMENTED** [review]
some minor comments on tests. otherwise LGTM

File: pkg/scheduler/internal/queue/scheduling_queue_test.go (hunk 7 above)
Comment (macsko) [reply pkg/scheduler/internal/queue/scheduling_queue_test.go:2852]: It's now duplicated with the "add two pod to activeQ and sort them by the timestamp" test above, because `heap.Add` and `heap.Update` were unified to `heap.AddAndUpdate`

File: pkg/scheduler/internal/heap/heap_test.go [hunk 8]
```diff
@@ -128,21 +131,34 @@ func TestHeap_Add(t *testing.T) {
...
+	// Update an item to a value that should push it to the head.
+	h.AddOrUpdate(mkHeapObj("baz", 0))
+	if h.data.queue[0] != "baz" || h.data.items["baz"].index != 0 {
+		t.Fatalf("expected baz to be at the head")
+	}
 	item, err := h.Pop()
-	if e, a := 10, item.(testHeapObject).val; err != nil || a != e {
+	if e, a := 0, item.val; err != nil || a != e {
 		t.Fatalf("expected %d, got %d", e, a)
 	}
```
[the remark this answers is not in the document]
Comment (macsko) [reply pkg/scheduler/internal/heap/heap_test.go:157]: I brought it back. Now `TestHeap_AddOrUpdate_Add` tests the add capabilities and `TestHeap_AddOrUpdate_Update` tests the update capabilities of `AddOrUpdate` method.

[the remark this answers is not in the document]
Comment (gabesaba) [reply pkg/scheduler/internal/heap/heap_test.go:159]: Could you revert the changes to `TestHeap_Delete` please?

[the remark this answers is not in the document]
Comment (macsko) [reply pkg/scheduler/internal/heap/heap_test.go:147]: Done, didn't notice them before

**Review by sanposhiho — APPROVED** [review]
/approve

Leave `/lgtm` to @gabesaba. Also please rebase upon the latest master. (will hit code-freeze anyway, though)
